### PR TITLE
Draft: Database redesign

### PR DESCRIPTION
**Pull request by Zenb0t**
## Description

This PR draft address issue https://github.com/Zenb0t/a3c1d7/issues/5
I added a Group model to the db models, to suggest a way to allow communication between more than 2 users, in groups of 3 or more. The following ERD diagram shows the new db structure:

Group Chat drawio

## Approach and thought process

The way this would work is every conversation is considered a group conversation, even between 2 users. All our logic can be the same, the difference would be only on how we present those in the frontend. Some high level requirements that would surface from these changes are:

- Users can create/own a(many) group(s)
- Users can invite others to a group
- User can leave a group
- Group owner have "admin" powers ( remove others, rename group)
- Multiple admins per group?
- Some form of "tag" to point to a certain user (like Slack or GitHub @)
- Mute a group conversation
- Get notified when invited to join

This feature would create the need for many frontend changes, including adding some sort of settings or preferences, be it by group or general, along with interfaces to create, join and leave groups. Additional filtering options would also be interesting.
For the backend, most of the group logic would live there. There would be more changes to the ERD as this matures for sure.
We would probably need to implement some sort of user role, and maybe have an extra table for it. As this also implies that the amount of messages will increase, we should think about implementing limits on how old messages can be retrieved on a first GET call.

One thing to note from the above ERD, the Conversation table is redundant, as we can connect Message straight to Group. I only kept it to avoid a big refactor of the app at this moment, and to not break functionality. This way we can initiate implementing it without breaking anything for our current users. I suggest creating an internal test server, where the team can use it to iron out any bugs before creating a beta feature. Of course, to be sure that it works properly before launch, a decent unit and integration test coverage would be very important, to allow CD/CI pipelines.

I'm sure that we could implement it with different approaches, but all of them seemed more complex from a database perspective. Trying to make conversations have multiple users would be a lot of legwork, as it introduces a few M:N relationships between Users, Conversations and Users, as many users can be in many conversations with many other users. Also, getting a read status would request a few queries at least, as we would need to transverse a few tables.  Taking away the relationship between conversation and user simplify the design from my point of view.

### CODE REVIEW

**Review by uimarshall — COMMENTED** [review]
Hi! @Zenb0t 
Thank you for your effort on this PR! so far 👍 especially for taking the time to give your thought process for the implementation.
I saw that you made a lot of changes to the codebase. This ticket is not to implement this full change to the codebase, but to help us plan out what that change would look like. Making changes to just the database models would have sufficed. 
I've left some additional comments on the code.
Also, here are a few additional questions:

- How would you update `socket event handlers` to reflect the schema change?
- How would you update `Redux state` to reflect the schema change?
- How would you update `route handlers` to reflect the schema change?
- Can you briefly describe the procedure to remove the deprecated models?
- For the migration plan, can you explain the step by step approach you would adopt to migrate the old data (present in the old database schema) to the new schema without disrupting service for clients?
- If we have multiple developers involved in this project, how can we split the task between them without introducing merge conflict?

Please append the answers to this PR and let us know on Slack when it's ready for a second review 🙂

**Review by andreluciani — APPROVED** [review]
Thanks for your answers!

> As far as I know, there is no Redux in this app. Could you clarify?

You're right, I think Marshall meant the React state, i.e., what changes would be needed to update the state correctly for the new feature.

> I would do something like /api/group:id , and would create a special route "conversation" to handle 1-1 chats.

Wouldn't the `/api/group:id` route be enough to handle 1-1 chats? What if a 1-1 chat becomes a group and vice-versa?

I'm approving this PR but please don't merge it since it's supposed to be a draft.

**Pull request closed without merging**
